misc/memory-connection-limits/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+        }
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ConnectionHandler = dummy::ConnectionHandler;
+    type ToSwarm = Void;
+
+    fn handle_pending_inbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: &Multiaddr,
+        _: &Multiaddr,
+    ) -> Result<(), ConnectionDenied> {
+        self.limits.check_limit(&mut self.mem_tracker)
+    }
+
+    fn handle_pending_outbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: Option<PeerId>,
+        _: &[Multiaddr],
+        _: Endpoint,
+    ) -> Result<Vec<Multiaddr>, ConnectionDenied> {
+        self.limits.check_limit(&mut self.mem_tracker)?;
+        Ok(vec![])
+    }
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: PeerId,
+        _: &Multiaddr,
+        _: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        Ok(dummy::ConnectionHandler)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _: ConnectionId,
+        _: PeerId,
+        _: &Multiaddr,
+        _: Endpoint,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        Ok(dummy::ConnectionHandler)
+    }
+
+    fn on_swarm_event(&mut self, _: FromSwarm<Self::ConnectionHandler>) {}
+
+    fn on_connection_handler_event(
+        &mut self,
+        _id: PeerId,
+        _: ConnectionId,
+        event: THandlerOutEvent<Self>,
+    ) {
+        void::unreachable(event)
+    }
+
+    fn poll(
+        &mut self,
+        _: &mut Context<'_>,
+        _: &mut impl PollParameters,
+    ) -> Poll<ToSwarm<Self::ToSwarm, THandlerInEvent<Self>>> {
+        Poll::Pending
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum ConnectionKind {
+    PendingIncoming,
+    PendingOutgoing,
+    EstablishedIncoming,
+    EstablishedOutgoing,
+    EstablishedPerPeer,
+    EstablishedTotal,
+}
+
+#[derive(Debug)]
+struct ProcessMemoryUsageTracker {
+    last_checked: Instant,
+    physical_bytes: usize,
+    virtual_bytes: usize,
+}
+
+impl ProcessMemoryUsageTracker {
+    fn new() -> Self {
+        let stats = memory_stats::memory_stats();
+        if stats.is_none() {
+            log::warn!("Failed to retrive process memory stats");
+        }
+        Self {
+            last_checked: Instant::now(),
+            physical_bytes: stats.map(|s| s.physical_mem).unwrap_or_default(),
+            virtual_bytes: stats
+                .map(|s: memory_stats::MemoryStats| s.virtual_mem)
+                .unwrap_or_default(),
+        }
+    }
+
+    fn need_refresh(&self) -> bool {
+        const PROCESS_MEMORY_USAGE_CHECK_INTERVAL: Duration = Duration::from_millis(1000);
+
+        self.last_checked + PROCESS_MEMORY_USAGE_CHECK_INTERVAL < Instant::now()
+    }
+
+    fn refresh(&mut self) {
+        if let Some(stats) = memory_stats::memory_stats() {
+            self.physical_bytes = stats.physical_mem;
+            self.virtual_bytes = stats.virtual_mem;
```

Comment:
Why not directly add them together here?

Reply:
When I run unit tests on windows sublinux 2 ubuntu, I see `virtual_mem` always being large, need to investigate whether it should be counted.